src/pal/src/thread/thread.cpp

```diff
@@ -1353,82 +1384,177 @@ GetThreadTimes(
         &pobjThread
         );
     
-	if (palError != NO_ERROR)
+    if (palError != NO_ERROR)
     {
         ASSERT("Unable to get thread data from handle %p"
               "thread\n", hThread);
         SetLastError(ERROR_INTERNAL_ERROR);
-        goto GetThreadTimesExit;
-	}	
+        goto SetTimesToZero;
+    }   
 
     pthrTarget->Lock(pthrCurrent);
-	
+    
     mach_port_t mhThread;
     mhThread = pthread_mach_thread_np(pthrTarget->GetPThreadSelf());
-	
-	kern_return_t status;
-	status = thread_info(
-	    mhThread, 
-		THREAD_BASIC_INFO, 
-		(thread_info_t)&resUsage, 
-		&resUsage_count);
-	
+    
+    kern_return_t status;
+    status = thread_info(
+        mhThread, 
+        THREAD_BASIC_INFO, 
+        (thread_info_t)&resUsage, 
+        &resUsage_count);
+
+    pthrTarget->Unlock(pthrCurrent);
+    
     if (status != KERN_SUCCESS)
-	{
+    {
         ASSERT("Unable to get resource usage information for the current "
               "thread\n");
         SetLastError(ERROR_INTERNAL_ERROR);
-        goto GetThreadTimesExit;
+        goto SetTimesToZero;
     }
+
+    /* Get the time of user mode execution, in nanoseconds */
+    calcTime = (__int64)resUsage.user_time.seconds * SECS_TO_NS;
+    calcTime += (__int64)resUsage.user_time.microseconds * USECS_TO_NS;
+    /* Assign the time into lpUserTime */
+    lpUserTime->dwLowDateTime = (DWORD)calcTime;
+    lpUserTime->dwHighDateTime = (DWORD)(calcTime >> 32);
+
+    /* Get the time of kernel mode execution, in nanoseconds */
+    calcTime = (__int64)resUsage.system_time.seconds * SECS_TO_NS;
+    calcTime += (__int64)resUsage.system_time.microseconds * USECS_TO_NS;
+    /* Assign the time into lpKernelTime */
+    lpKernelTime->dwLowDateTime = (DWORD)calcTime;
+    lpKernelTime->dwHighDateTime = (DWORD)(calcTime >> 32);
+
+    retval = TRUE;
+
+    goto GetThreadTimesInternalExit;
+
+#else //HAVE_MACH_THREADS
+
+    if (!IsCurrentThread(hThread))
+    {
+        //Set all times to 0 if this isn't current thread
+        goto SetTimesToZero;
+    }
+
+#if HAVE_CLOCK_THREAD_CPUTIME
+    struct timespec ts;
+    if (clock_gettime(CLOCK_THREAD_CPUTIME_ID, &ts) != 0)
+    {
+        ASSERT("clock_gettime(CLOCK_THREAD_CPUTIME_ID) failed; errno is %d (%s)\n", errno, strerror(errno));
+        goto SetTimesToZero;
+    }
+
+    if (lpUserTime)
```

Comment:
Would you suggest moving isCurrentThread out to GetThreadTimes then, before the call to GetThreadTimesInternal? Otherwise it will never get called at all.

From: Sergiy Kuryata [mailto:notifications@github.com]
Sent: Tuesday, October 6, 2015 9:32 AM
To: dotnet/coreclr
Cc: Will Godbe
Subject: Re: [coreclr] Implemented QueryThreadCycleTime (time.cpp) & GetThreadTimes (thread.… (#1671)


In src/pal/src/thread/thread.cpp<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fdotnet%2fcoreclr%2fpull%2f1671%23discussion_r41288147&data=01%7c01%7cwigodbe%40microsoft.com%7cbf4204e4f0434177a63f08d2ce6b9bae%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=9D%2b4lLjKiGUnu%2fnMMFP5n%2f95fSV7xsCDOY88pmuuArQ%3d>:

> +

> +    if (!IsCurrentThread(hThread))

> +    {

> +        //Set all times to 0 if this isn't current thread

> +        goto SetTimesToZero;

> +    }

> +

> +#if HAVE_CLOCK_THREAD_CPUTIME

> +    struct timespec ts;

> +    if (clock_gettime(CLOCK_THREAD_CPUTIME_ID, &ts) != 0)

> +    {

> +        ASSERT("clock_gettime(CLOCK_THREAD_CPUTIME_ID) failed; errno is %d (%s)\n", errno, strerror(errno));

> +        goto SetTimesToZero;

> +    }

> +

> +    if (lpUserTime)

Since GetThreadTimesInternal is an internal function, we don't need this check. We should assert that it is not NULL instead. The caller must pass a valid parameter.

—
Reply to this email directly or view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fdotnet%2fcoreclr%2fpull%2f1671%2ffiles%23r41288147&data=01%7c01%7cwigodbe%40microsoft.com%7cbf4204e4f0434177a63f08d2ce6b9bae%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=K5jy0tkfBH%2bYWBIBx9wcyGnj8uMyz2F3Ld%2f5KTShP%2bw%3d>.